// @thirdweb-dev/wallets/evm/wallets/signer/dist/thirdweb-dev-wallets-evm-wallets-signer.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var walletIds = require('../../../../dist/walletIds-0e1ebc43.cjs.dev.js');
var evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase = require('../../../../dist/base-7769b2e0.cjs.dev.js');
var chains = require('@thirdweb-dev/chains');
require('../../abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.cjs.dev.js');
require('../../../../dist/defineProperty-982e7acf.cjs.dev.js');
require('ethers');
require('eventemitter3');
require('@thirdweb-dev/sdk');
require('../../../../dist/headers-e56e888d.cjs.dev.js');

// eslint-disable-next-line @typescript-eslint/ban-types

/**
 * @internal
 */
class SignerWallet extends evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.AbstractClientWallet {
  constructor(options) {
    super("signerWallet", options);
    if (options.clientId && options.chain) {
      options.chain = chains.updateChainRPCs(options.chain, options.clientId);
    }
    this.options = options;
    this.signer = options.signer;
    this._storage = options?.storage || evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.createAsyncLocalStorage(walletIds.walletIds.localWallet);
  }
  async getConnector() {
    if (!this.connector) {
      const {
        SignerConnector
      } = await Promise.resolve().then(function () { return require('../../../connectors/signer/dist/thirdweb-dev-wallets-evm-connectors-signer.cjs.dev.js'); });
      if (!this.signer) {
        this.signer = this.options.signer;
      }
      const defaults = (this.options.chain ? [...chains.defaultChains, this.options.chain] : chains.defaultChains).map(c => chains.updateChainRPCs(c, this.options.clientId));
      this.connector = new SignerConnector({
        chain: this.options.chain || chains.updateChainRPCs(chains.Ethereum, this.options.clientId),
        signer: this.signer,
        chains: this.chains || defaults,
        clientId: this.options.clientId,
        secretKey: this.options.secretKey
      });
    }
    return this.connector;
  }
}

exports.SignerWallet = SignerWallet;
